fe/fe-core/src/main/java/com/starrocks/qe/scheduler/dag/CaptureVersionFragmentBuilder.java

```diff
@@ -0,0 +1,108 @@
+    private final List<ExecutionFragment> fragments;
+
+    public CaptureVersionFragmentBuilder(List<ExecutionFragment> fragments) {
+        this.fragments = fragments;
+    }
+
+    public ExecutionFragment build(ExecutionDAG dag) {
+        Map<ComputeNode, List<TScanRangeParams>> workerId2ScanRanges = Maps.newHashMap();
+        int id = 0;
+
+        for (ExecutionFragment fragment : fragments) {
+            final PlanFragmentId fragmentId = fragment.getFragmentId();
+            for (FragmentInstance instance : fragment.getInstances()) {
+                final Set<Integer> localNativeScanNodeIds = instance.getExecFragment().getScanNodes().stream()
+                        .filter(ScanNode::isLocalNativeTable).map(node -> node.getId().asInt()).collect(
+                                Collectors.toSet());
+                final ComputeNode worker = instance.getWorker();
+                final Map<Integer, List<TScanRangeParams>> node2ScanRanges = instance.getNode2ScanRanges();
+                final Map<Integer, Map<Integer, List<TScanRangeParams>>> node2DriverSeqToScanRanges =
+                        instance.getNode2DriverSeqToScanRanges();
+
+                // collect olap per driver scan ranges
+                final Stream<TScanRangeParams> tabletStream = node2DriverSeqToScanRanges.entrySet().stream()
+                        .filter(e -> localNativeScanNodeIds.contains(e.getKey()))
+                        .map(k -> k.getValue().values()).flatMap(Collection::stream).flatMap(Collection::stream);
+
+                // collect olap normal scan ranges
+                final Stream<TScanRangeParams> nodeStream =
+                        node2ScanRanges.entrySet().stream().filter(e -> localNativeScanNodeIds.contains(e.getKey()))
+                                .map(Map.Entry::getValue).flatMap(Collection::stream);
+
+                final List<TScanRangeParams> instanceScanRanges =
+                        Stream.concat(tabletStream, nodeStream).filter(r -> r.scan_range.isSetInternal_scan_range())
+                                .collect(Collectors.toList());
+
+                if (instanceScanRanges.isEmpty()) {
+                    continue;
+                }
+
+                final List<TScanRangeParams> workerIdScanRanges =
+                        workerId2ScanRanges.computeIfAbsent(worker, k -> Lists.newArrayList());
+                workerIdScanRanges.addAll(instanceScanRanges);
+            }
+            id = Math.max(id, fragmentId.asInt());
+        }
+
+        if (!workerId2ScanRanges.isEmpty()) {
+            final PlanFragmentId captureVersionFragmentId = new PlanFragmentId(id + 1);
+            final int fid = captureVersionFragmentId.asInt();
+            // Just get a tuple id.
+            final ArrayList<TupleId> tupleIds = fragments.get(0).getPlanFragment().getPlanRoot().getTupleIds();
+            PlanNode dummyNode = new CaptureVersionNode(PlanNodeId.DUMMY_PLAN_NODE_ID, tupleIds);
```

Comment:
why not put the CaptureVersionNode->BlockHoleTableSink's construction work into the PlanFragmentBuilder?

Visible versions of  all involved tablets of the OlapScanNodes would be captured at first via this Fragment  CaptureVersionNode->BlockHoleTableSink?

Do there exist optimized scenarios that some of OlapScanNodes should be captured, while some of them should not be captured?

For an example,  the left side of left-semi/anti-join seems should capture the versions,  while the right side need not? if we put  CaptureVersionNode->BlockHoleTableSink's construction into PlanFragmentBuilder, we can  control version capturing highly precisely, we can ignore this  PlanFragment in explain's output.

Reply:
> why not put the CaptureVersionNode->BlockHoleTableSink's construction work into the PlanFragmentBuilder?

I will try it.

>Visible versions of all involved tablets of the OlapScanNodes would be captured at first via this Fragment 
CaptureVersionNode->BlockHoleTableSink?

yes

> Do there exist optimized scenarios that some of OlapScanNodes should be captured, while some of them should not be captured?

The first scan node to be delivered does not have to capture the version. 
In the all-at-once implementation we assume that the scan range for delivering a batch is fast.
However, a simple join build side of a phased schedule can be very slow, and the probe side must capture the rowset version. A possible optimization would be to use the tablets involved in the undelivered scan nodes after more than N seconds if the delivery is not finished. One possible optimization is to capture the tablet involved in the undelivered scan node after more than N seconds if the delivery is not finished.


> For an example, the left side of left-semi/anti-join seems should capture the versions, while the right side need not? if we put CaptureVersionNode->BlockHoleTableSink's construction into PlanFragmentBuilder, we can control version capturing highly precisely, we can ignore this PlanFragment in explain's output.

The planfragment Builder stage doesn't yet perceive which fragments are the first to be delivered.